staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,616 @@
+//  request and afterExecution is irrelevant.
+func (qs *queueSetImpl) Wait(ctx context.Context, hashValue uint64) (tryAnother, execute bool, afterExecution func()) {
+	// TODO(aaron-prindle) verify locking is correct
+
+	qs.lockAndSyncTime()
+	// TODO(aaron-prindle) verify and test quiescent
+	// A call to Wait while the system is quiescent will be rebuffed by
+	// returning `tryAnother=true`.
+	if qs.emptyHandler != nil {
+		qs.lock.Unlock()
+		return true, false, func() {}
+	}
+
+	// ========================================================================
+	// Step 1:
+	// 1) Start with shuffle sharding, to pick a queue.
+	// 2) Reject old requests that have been waiting too long
+	// 3) Reject current request if there is not enough concurrency shares and
+	// we are at max queue length
+	// 4) If not rejected, create a packet and enqueue
+	req := qs.timeoutOldRequestsAndRejectOrEnqueue(hashValue)
+	// req == nil means that the request was rejected - no remaining
+	// concurrency shares and at max queue length already
+	if req == nil {
+		qs.lock.Unlock()
+		metrics.AddReject(qs.name, "queue-full")
+		return false, false, func() {}
+	}
+	// ========================================================================
+
+	// ------------------------------------------------------------------------
+	// Step 2:
+	// 1) The next step is to invoke the method that dequeues as much as possible.
+
+	// This method runs a loop, as long as there
+	// are non-empty queues and the number currently executing is less than the
+	// assured concurrency value.  The body of the loop uses the fair queuing
+	// technique to pick a queue, dequeue the request at the head of that
+	// queue, increment the count of the number executing, and send true to
+	// the request's channel.
+	qs.dequeueWithChannelAsMuchAsPossible()
+	// ------------------------------------------------------------------------
+	qs.lock.Unlock()
+
+	// ************************************************************************
+	// Step 3:
+	// After that method finishes its loop and returns, the final step in Wait
+	// is to `select` (wait) on a message from the enqueud packet's channel
+	// and return appropriately.
+	qs.counter.Add(-1)
+
+	select {
+	case execute := <-req.DequeueChannel:
+		if execute {
+			// execute the request
+			return false, true, func() {
+				qs.finishRequestAndDequeueWithChannelAsMuchAsPossible(req)
+			}
+
+		}
+		klog.V(5).Infof("request timed out after being enqueued\n")
+		metrics.AddReject(qs.name, "time-out")
+		return false, false, func() {}
+	case <-ctx.Done():
+		klog.V(5).Infof("request timed out\n")
+		if req.Enqueued {
+			// remove the request from the queue as it has timed out
+			func() {
+				qs.lockAndSyncTime()
+				defer qs.lock.Unlock()
+				for i := range req.Queue.Requests {
+					if req == req.Queue.Requests[i] {
+						req.Queue.Requests = removeRequest(req.Queue.Requests, i)
+						break
+					}
+				}
+			}()
+		} else {
+			qs.counter.Add(-1)
```

Comment:
Why?

Reply:
(Supposing the test of `req.Enqueued` gets moved under the lock)

> // In this case we know that `req` was in the queue earlier and is not there now;
> // that must be because another goroutine has dequeued `req` and called
> // `qs.counter.Add(1)` in anticipation of triggering the other arm of this select.
> // Decrement the counter because that other arm is not actually going to fire.